sql-plugin/src/main/340+/scala/org/apache/spark/sql/execution/datasources/GpuWriteFiles.scala

```diff
@@ -0,0 +1,175 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources
+
+import java.util.Date
+
+import com.nvidia.spark.rapids.{DataFromReplacementRule, GpuExec, RapidsConf, RapidsMeta, SparkPlanMeta}
+import com.nvidia.spark.rapids.shims.ShimUnaryExecNode
+
+import org.apache.spark.{SparkException, TaskContext}
+import org.apache.spark.internal.io.{FileCommitProtocol, SparkHadoopWriterUtils}
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
+import org.apache.spark.sql.catalyst.catalog.CatalogTypes.TablePartitionSpec
+import org.apache.spark.sql.catalyst.expressions.{Attribute, SortOrder}
+import org.apache.spark.sql.connector.write.WriterCommitMessage
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.rapids.{GpuFileFormatWriter, GpuWriteJobDescription}
+import org.apache.spark.sql.rapids.GpuFileFormatWriter.GpuConcurrentOutputWriterSpec
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * The write files spec holds all information of [[V1WriteCommand]] if its provider is
+ * [[FileFormat]].
+ */
+case class GpuWriteFilesSpec(
+    description: GpuWriteJobDescription,
+    committer: FileCommitProtocol,
+    concurrentOutputWriterSpecFunc: SparkPlan => Option[GpuConcurrentOutputWriterSpec])
+
+class GpuWriteFilesMeta(
+    writeFilesExec: WriteFilesExec,
+    conf: RapidsConf,
+    parent: Option[RapidsMeta[_, _, _]],
+    rule: DataFromReplacementRule)
+    extends SparkPlanMeta[WriteFilesExec](writeFilesExec, conf, parent, rule) {
+
+  override def convertToGpu(): GpuExec = {
+    // WriteFilesExec only has one child
+    assert(childPlans.size == 1)
+
+    val child = childPlans.head.convertIfNeeded()
+    GpuWriteFilesExec(
+      child,
+      writeFilesExec.fileFormat,
+      writeFilesExec.partitionColumns,
+      writeFilesExec.bucketSpec,
+      writeFilesExec.options,
+      writeFilesExec.staticPartitions
+    )
+  }
+}
+
+/**
+ * Responsible for writing files.
+ */
+case class GpuWriteFilesExec(
+    child: SparkPlan,
+    fileFormat: FileFormat,
+    partitionColumns: Seq[Attribute],
+    bucketSpec: Option[BucketSpec],
+    options: Map[String, String],
+    staticPartitions: TablePartitionSpec) extends ShimUnaryExecNode with GpuExec {
+
+  override def output: Seq[Attribute] = Seq.empty
+
+  /**
+   * Cpu version for SparkPlan.executeWrite, just throw an exception
+   */
+  override protected def doExecuteWrite(writeFilesSpec: WriteFilesSpec): RDD[WriterCommitMessage] =
+    throw new UnsupportedOperationException(
+      s"${getClass.getCanonicalName} does not support row-based execution")
```

Comment:
Why would we throw if this is called?  This is not necessarily returning rows, it's returning the result of a commit, and we're using the same committer as the CPU does.  Unless I'm missing something, `doExecuteWrite` does not require row-based execution of the node.

Reply:
Renamed

Reply:
I was very confused how a "rename" fixes my original comment, but now I better understand what this is doing.  The key is `WriteFilesSpec` vs `GpuWriteFilesSpec`.  It's a subtle difference when reading, so I think we should go back to `executeColumnarWrite` to make this more clear.  Sorry I misread it originally.

Reply:
Done